### ISSUE TITLE
[sweep] function= 에 들어가는 함수에 매개변수가 있다면?

### ISSUE THREAD

**Issue by soohi0** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#68]
https://github.com/boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13/blob/8d4cdef1b25417653ca568b0af500131ab32160e/hyper_tune_train.py#L196

이거때문에 꽤나 고생한게 기억이 나네요, 
function 매개변수에 함수 이름만 적어야 한다는 사실을 몰랐기 때문입니다.

그런데 function에 넣을 함수자체에 매개변수가 있을 경우엔.. 아무리 찾아봐도 함수에 매개변수가 있는 함수를 function으로 넣은 예시를 못봐서 질문을 남깁니다.
기존 코드에선 `main(config)` 였는데 제가 매개변수를 따로 넘겨줄 수 있는 방법을 찾아보다가 못 찾아서 우선은  config 불러오는 부분은 main 내부에 넣고 main 매개변수를 아예 없앤 채로 function에 main 이름만 넘겨준 상태입니다. 

`function=main(config)` 이렇게 하면 에러가 발생합니다.😇

main 의 매개변수, 즉 config를 wandb.agent 로 넘겨줄 수 있는 방법이 있을까요? 
document 아무리 찾아봐도 보이지 않습니다...

**Comment by intrandom5** [comment]
이건 내가 이번에 코드 짜면서 해결했던 이슈 중 하나랑 비슷한 문제인 것 같아서 내 사례를 소개해줄게.
[문제점]
collate_fn에 pad_id 인자를 추가로 전달하고 싶음. 그러나 dataloader는 collate_fn을 아래와 같이 작성해서 인자를 따로 넘길 방법이 없음.
>
    train_loader = DataLoader(train_datasets, batch_size=64, shuffle=True, collate_fn=collate_fn)

collate_fn은 다음과 같이 작성됨.
>
    def collate_fn(batch):
        s1 = []
        labels = []
        for b in batch:
            data, label, aux = b
            s1.append(data)
            labels.append(label)
            s1_batch = pad_sequence(s1, batch_first=True, padding_value=0)
            return s1_batch.long(), torch.LongTensor(labels)
함수로 구성하면 이렇게 batch밖에 인자로 전달할 수 없지만, 이를 객체화 하면 아래와 같이 추가 인자를 전달할 수 있다.
>
    class Collate_fn(object):
        def __init__(self, pad_id):
            self.pad_id = pad_id

        def __call__(self, batch):
            s1 = []
            labels = []
            for b in batch:
                data, label, aux = b
                s1.append(data)
                labels.append(label)
                s1_batch = pad_sequence(s1, batch_first=True, padding_value=self.pad_id)
                return s1_batch.long(), torch.LongTensor(labels)

    collate_fn = Collate_fn(pad_id=0)

    train_loader = DataLoader(train_datasets, batch_size=64, shuffle=True, collate_fn=collate_fn)

sweep도 마찬가지로 적용 가능하다고 생각.
여기선 "main"함수라서 객체화가 적합한 방법인지는 조금 의문이긴 한데, 해결은 가능할 것임.
>
    class Main(object):
        def __init__(self, config):
            self.config = config
        def __call__():
            main 코드

    wandb.agent(sweep_id=sweep_id, function = main, count=20) 

근데 main은 특수한 함수라서 main을 저렇게 하기보다는 training 과정을 담고 있는 부분만 코드로 따로 빼서 인자로 전달하는게 맞는 것 같습니다. 아무튼 저렇게 해결할 수 있다

**Comment by soohi0** [comment]
.

**Comment by soohi0** [comment]
.
